File: scripts/build-canonical-direct.py
import base64, gzip, json, pathlib, re

ROOT=pathlib.Path(__file__).resolve().parents[1]
V1=ROOT/'src'/'v1'
OUT=V1/'canonical-history-direct.ts'
VENUE_SLUG={'札幌':'sapporo','函館':'hakodate','福島':'fukushima','新潟':'niigata','東京':'tokyo','中山':'nakayama','中京':'chukyo','京都':'kyoto','阪神':'hanshin','小倉':'kokura'}

def selected(year:int):
    text=(V1/f'frozen-selected-{year}.ts').read_text(encoding='utf-8')
    body=text.split('=',1)[1].rsplit(';',1)[0]
    return json.loads(body)

def bin_text(i:int):
    text=(V1/'canonical-history-data'/f'bin-{i:02d}.ts').read_text(encoding='utf-8')
    m=re.search(r'export default\s+(["\'])(.*?)\1\s*;?\s*$',text,re.S)
    if not m: raise RuntimeError(f'BAD_BIN:{i}')
    return bytes(m.group(2),'utf-8').decode('unicode_escape')

def read_varint(data:bytes,pos:int):
    value=0; shift=0
    while pos<len(data):
        b=data[pos]; pos+=1; value|=(b&127)<<shift
        if b<128: return value,pos
        shift+=7
        if shift>28: raise RuntimeError('VARINT_TOO_LARGE')
    raise RuntimeError('TRUNCATED')

ids=[]
for year in (2024,2025,2026):
    for key,value in selected(year).items():
        date,venue=key.split('|',1)
        if date>'2026-08-02': continue
        slug=VENUE_SLUG[venue]
        for race_no in value.split('.'):
            ids.append(f'{date}-{slug}-{int(race_no):02d}')
if len(ids)!=3210: raise RuntimeError(f'RACE_COUNT:{len(ids)}')

encoded=''.join(bin_text(i) for i in range(4))
data=gzip.decompress(base64.b64decode(encoded))
pos=0
count,pos=read_varint(data,pos)
if count!=len(ids): raise RuntimeError(f'ARCHIVE_COUNT:{count}/{len(ids)}')
out={}; ticket_count=0
for race_id in ids:
    n,pos=read_varint(data,pos); rows=[]
    for _ in range(n):
        bet,pos=read_varint(data,pos); combo,pos=read_varint(data,pos); obin,pos=read_varint(data,pos)
        rows.append([bet,combo,obin,0])
    winners,pos=read_varint(data,pos)
    for _ in range(winners):
        idx,pos=read_varint(data,pos); payout,pos=read_varint(data,pos); rows[idx][3]=payout
    ticket_count+=n; out[race_id]=rows
if pos!=len(data) or ticket_count!=17735: raise RuntimeError(f'INTEGRITY:{pos}/{len(data)}/{ticket_count}')

payload=json.dumps(out,ensure_ascii=False,separators=(',',':'))
OUT.write_text('export const DIRECT_CANONICAL_HISTORY:Record<string,number[][]>='+payload+';\n',encoding='utf-8')
print(json.dumps({'races':len(out),'tickets':ticket_count,'bytes':OUT.stat().st_size},ensure_ascii=False))

File: scripts/build-canonical-direct.mjs
import fs from 'node:fs';
import path from 'node:path';
import zlib from 'node:zlib';

const ROOT=process.cwd();
const V1=path.join(ROOT,'src','v1');
const OUT=path.join(V1,'canonical-history-direct.ts');
const VENUE_SLUG={札幌:'sapporo',函館:'hakodate',福島:'fukushima',新潟:'niigata',東京:'tokyo',中山:'nakayama',中京:'chukyo',京都:'kyoto',阪神:'hanshin',小倉:'kokura'};

function defaultString(i){
  const src=fs.readFileSync(path.join(V1,'canonical-history-data',`bin-${String(i).padStart(2,'0')}.ts`),'utf8').trim();
  const literal=src.replace(/^export\s+default\s+/,'').replace(/;\s*$/,'');
  const value=JSON.parse(literal);
  if(typeof value!=='string')throw new Error(`BAD_BIN_${i}`);
  return value;
}
function selected(year){
  const src=fs.readFileSync(path.join(V1,`frozen-selected-${year}.ts`),'utf8');
  const m=src.match(/=\s*(\{[\s\S]*\})\s*;\s*$/);
  if(!m)throw new Error(`BAD_SELECTED_${year}`);
  return JSON.parse(m[1]);
}
function readVarint(buf,state){
  let value=0,shift=0;
  while(state.i<buf.length){const b=buf[state.i++];value|=(b&127)<<shift;if(b<128)return value;shift+=7;if(shift>28)throw new Error('VARINT_TOO_LARGE');}
  throw new Error('TRUNCATED');
}

const ids=[];
for(const year of [2024,2025,2026]){
  for(const [key,value] of Object.entries(selected(year))){
    const [date,venue]=key.split('|');if(date>'2026-08-02')continue;
    const slug=VENUE_SLUG[venue];if(!slug)throw new Error(`UNKNOWN_VENUE:${venue}`);
    for(const raceNo of String(value).split('.'))ids.push(`${date}-${slug}-${String(Number(raceNo)).padStart(2,'0')}`);
  }
}
if(ids.length!==3210)throw new Error(`RACE_COUNT:${ids.length}`);

const chunks=[0,1,2,3].map(defaultString);
const encoded=chunks.join('');
const compressed=Buffer.from(encoded,'base64');
const data=zlib.gunzipSync(compressed);
const state={i:0};
const count=readVarint(data,state);if(count!==ids.length)throw new Error(`ARCHIVE_COUNT:${count}/${ids.length}`);
const out={};let ticketCount=0;
for(const raceId of ids){
  const n=readVarint(data,state),rows=[];
  for(let i=0;i<n;i++)rows.push([readVarint(data,state),readVarint(data,state),readVarint(data,state),0]);
  const winners=readVarint(data,state);
  for(let i=0;i<winners;i++){const idx=readVarint(data,state),payout=readVarint(data,state);if(!rows[idx])throw new Error('BAD_WINNER_INDEX');rows[idx][3]=payout;}
  ticketCount+=n;out[raceId]=rows;
}
if(state.i!==data.length||ticketCount!==17735)throw new Error(`INTEGRITY:${state.i}/${data.length}/${ticketCount}`);
fs.writeFileSync(OUT,`export const DIRECT_CANONICAL_HISTORY:Record<string,number[][]>=${JSON.stringify(out)};\n`);
console.log(JSON.stringify({chunks:chunks.map(x=>x.length),encoded:encoded.length,compressed:compressed.length,uncompressed:data.length,races:Object.keys(out).length,tickets:ticketCount,bytes:fs.statSync(OUT).size}));
